File: Medium/3121_Count_the_Number_of_Special_Characters_II/Solution.cpp
#include "../../devlibs/cpp/cpphelper.h"

using namespace std;


// NOTE:
// For LeetCode submission, copy only the `class Solution` part.


class Solution {
public:
    int numberOfSpecialChars(string word) {
        vector<int> firstUpper(26, -1);
        vector<int> lastLower(26, -1);

        for (int i = 0; i < word.size(); i++) {
            char c = word[i];

            if (islower(c)) {
                lastLower[c - 'a'] = i;
            } 
            else {
                if (firstUpper[c - 'A'] == -1) {
                    firstUpper[c - 'A'] = i;
                }
            }
        }

        int count = 0;

        for (int i = 0; i < 26; i++) {
            if (lastLower[i] != -1 && firstUpper[i] != -1 && lastLower[i] < firstUpper[i]) {
                count++;
            }
        }

        return count;
    }
};


int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    Solution solution;

    // test cases 1
    cout << solution.numberOfSpecialChars("aaAbcBC") << endl;

    // test cases 2
    cout << solution.numberOfSpecialChars("abc") << endl;

    // test cases 3
    cout << solution.numberOfSpecialChars("AbBCab") << endl;

    return 0;
}

File: Medium/3121_Count_the_Number_of_Special_Characters_II/Solution.py
from typing import List, Optional


class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        firstUpper = [-1] * 26
        lastLower = [-1] * 26

        for i in range(len(word)):
            c = word[i]

            if (c.islower()):
                lastLower[ord(c) - ord('a')] = i
            
            else:
                if (firstUpper[ord(c) - ord('A')] == -1):
                    firstUpper[ord(c) - ord('A')] = i
        
        count = 0

        for i in range(26):
            if (lastLower[i] != -1 and firstUpper[i] != -1 and lastLower[i] < firstUpper[i]):
                count += 1
        
        return count
            


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    print(solution.numberOfSpecialChars("aaAbcBC"))

    # test cases 2
    print(solution.numberOfSpecialChars("abc"))

    # test cases 3
    print(solution.numberOfSpecialChars("AbBCab"))
    
    